### PR TITLE
Adds metrics plan for add-on and recommendation server.

### PR DESCRIPTION
**Pull request by chuckharmston**
Nicer-reading version here: https://github.com/chuckharmston/universal-search/blob/18-metrics-plan/docs/metrics.md

r? @6a68 @clouserw @nchapman

Connected to #18, connected to #35.

### CODE REVIEW

File: metrics.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+- The HTTP status code of the response.
+- The enhancers applied to that query (i.e. `recommendation.enhancements.keys()`)
+
+We do not wish to record individual users browsing habits; we're interested in aggregations of user behavior. Thus, we will not record any queries that:
+
+- Are longer than 20 characters.
+- Appear to begin with a protocol (`^[^\s]+\:\S`).
+- Appear to be a hostname (`^[^\s]+\.\S`)
+
+Additionally, on a TBD schedule we will discard outliers:
```
Comment (chuckharmston) [inline metrics.md]: I'm not sure if/how this is possible in the Heka/Redshift flow.

File: metrics.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+## Client metrics
+
+Metrics gathered from the Universal Search add-on will be reported to the Test Pilot add-on, which will augment that information and relay it to Firefox's telemetry system via `submitExternalPing`. This is not currently implemented on the Test Pilot side; those efforts are encompassed by [issue #234](https://github.com/mozilla/testpilot/issues/234).
+
+
+### Events
+
+Universal Search will record and report each of the following events:
+
+- User selects a recommendation.
```
Comment (ghost) [inline metrics.md]: You have these broken down by event, but you have a lot of duplication (eg. time from start to end).  Wouldn't "event" just be a type identifier?  So you'd record something like:

```
us.totaltime
us.querylength
us.eventtype
us.enhancers (if applicable)
us.datatype (if applicable)
us.levenstein (if applicable)
```

and then when you visualize you'd break it down into whatever events you wanted?  I know this is just the same thing you're saying here, but I'm trying to picture this from the POV of recording the data.  (that will help build the right structures to record the data)

File: metrics.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+Once a day, Universal Search will report an aggregation of activity from the previous day. That data will include:
+
+- The number of searches in the previous day.
+- The number of those searches that ended in each of the following ways:
+	- By visiting a recommendation.
+	- By entering and visiting a URL directly into the URL bar.
+	- By entering and visiting a search query into the URL bar.
+	- By using a suggested search term.
+	- By visiting a bookmark.
```
Comment (ghost) [inline metrics.md]: Is this (and the history) meaning a bookmark that shows up in the URL bar, or would clicking a bookmark in the bookmark menu do the same thing (that is to say, clicking a link with no interaction with universal search at all)

File: metrics.md (hunk 1 above)
Comment (ghost) [inline metrics.md]: Me either.  We'll have to ask.

File: metrics.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
+# Universal Search Metrics
```
Comment (ghost) [inline metrics.md]: This is a really good start, thanks Chuck.  Can you identify what the most important KPI is?  (probably something with engagement ratio?)

File: metrics.md (hunk 2 above)
Comment (chuckharmston) [inline metrics.md]: Yep, that's likely how it'd be implemented. I was trying to sketch the telemetry plan agnostic of that, except where it would dictate the collection strategy (on the server).

File: metrics.md (hunk 1 above)
Comment (chuckharmston) [inline metrics.md]: Any idea who the correct person is to ask about this?

File: metrics.md (hunk 3 above)
Comment (chuckharmston) [inline metrics.md]: Naively, this would specifically mean by clicking on a bookmark in the awesome bar. This is somewhat indirectly implied by the use of "those searches", with "search" being defined at the top.

File: metrics.md (hunk 4 above)
Comment (chuckharmston) [inline metrics.md]: @clouserw Is the official Mozilla Engagement Ratio formula finished? Last I saw it was a draft in a Google Doc somewhere.

Comment (ghost) [inline metrics.md]: nope, still draft as far as I know.  The discussion for Fx is happening in https://bugzilla.mozilla.org/show_bug.cgi?id=1240849

File: metrics.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+	- Time from search start to search completion.
+	- Length of query.
+- Use selects an item from history.
+	- Time from search start to search completion.
+	- Length of query.
+
+
+### Daily ping
+
+Once a day, Universal Search will report an aggregation of activity from the previous day. That data will include:
```
Comment (jaredhirsch) [inline metrics.md]: What if the experiments just send metrics to test pilot as they are generated? That lets experiments avoid the work of implementing web storage solutions to hold metrics data, and lets the test pilot addon and server negotiate an update frequency.

File: metrics.md (hunk 1 above)
Comment (jaredhirsch) [inline metrics.md]: I'd ask in #datapipeline

Comment (clouserw) [inline metrics.md]: I talked to Rob.  Sounds like it's fairly trivial to purge data from Redshift, and slightly more difficult to purge from S3, but we could put a retention policy on S3 since it's just a stop on the way to Redshift anyway.

tl;dr this should be possible.

File: metrics.md (hunk 2 above)
Comment (nchapman) [inline metrics.md]: Yeah I agree that it could be a single event like selected/navigated/closed (whatever the right way to think about this) with these attributes attached to it.

File: metrics.md [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+	- Time from search start to search completion.
+	- Length of query.
+	- Levenstein distance between query and completed search term.
+- User selects a bookmark.
+	- Time from search start to search completion.
+	- Length of query.
+- Use selects an item from history.
+	- Time from search start to search completion.
+	- Length of query.
+
```
Comment (nchapman) [inline metrics.md]: Probably worth recording when the user chooses nothing. This will be especially important when we attempt to answer questions in our recommendations such as with a Weather card.

File: metrics.md (hunk 5 above)
Comment (nchapman) [inline metrics.md]: Seems like a great idea.

File: metrics.md (hunk 1 above)
Comment (nchapman) [inline metrics.md]: Marshall would have some good input on this as well from the policy side.

File: metrics.md (hunk 4 above)
Comment (nchapman) [inline metrics.md]: I agree this is a great start! I think we will end up reporting the engagement ratio for all Test Pilot tests "automatically". The metrics that I think are most important for this test specifically are the click through rates of the recommendations and retention. I'm working on a summary doc right now to capture these. We can integrate into this doc as soon as that's done.

File: metrics.md (hunk 6 above)
Comment (jaredhirsch) [inline metrics.md]: We can break down 'user chooses nothing' in a few ways:
- did the user hit `Escape`, `LeftArrow`, or `RightArrow` to dismiss the popup?
- did the user switch tabs or windows?
- did the user switch focus away from the Firefox application altogether?

Also, I think it's worth recording whether a user's selection involves a key or mouse event.

**Comment by jaredhirsch** [comment]
@chuckharmston Looks like a good start, I added a few comments inline. Thanks for driving this.

Data retention and access controls aren't discussed, but probably should be. I could see legal/privacy making those decisions for us, but it'd be good to mention them here, since users might think omitting those details implies that we don't have rules in place.

Feel free to reassign to me if you want to do a second round of feedback.

File: metrics.md (hunk 5 above)
Comment (chuckharmston) [inline metrics.md]: This was written assuming that we didn't have access to a UID, which would have make any sort of proportional reporting (i.e. x% of the time, y outcome) impossible. Turns out, Test Pilot will be giving us that, so we should send them as they're generated.

File: metrics.md (hunk 2 above)
Comment (chuckharmston) [inline metrics.md]: Agreed. As I said above, I'm avoiding jumping to implementation until we have reporting APIs to develop against.

File: metrics.md (hunk 6 above)
Comment (chuckharmston) [inline metrics.md]: IRC chat with @6a68: it's probably enough to know that it closed, and not how they closed it.

**Comment by chuckharmston** [comment]
Updated to address comments and add KPIs with some visualization (thanks, Nick!).

r? @6a68 

(Feel free to merge if you're happy where it's at; I think it's ready.)

File: docs/metrics.md [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+	- Time from search start to search completion.
+	- Length of query.
+	- Levenstein distance between query and completed search term.
+- User selects a bookmark.
+	- Time from search start to search completion.
+	- Length of query.
+- Use selects an item from history.
+	- Time from search start to search completion.
+	- Length of query.
+- The user terminates a search.
```
Comment (clouserw) [inline docs/metrics.md:41]: I'd like to see the actual JSON object (as part of this documentation) you'd be sending as the payload in submitExternalPing() to track all of these.

File: docs/metrics.md [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+To gain a better understanding of how users make use of the URL bar, the recommendation server will maintain logs of queries passed to it. For each request, we will record:
+
+- The query string itself.
+- The HTTP status code of the response.
+- The enhancers applied to that query (i.e. `recommendation.enhancements.keys()`)
+
+We do not wish to record individual users browsing habits; we're interested in aggregations of user behavior. Thus, we will not record any queries that:
+
+- Are longer than 20 characters.
+- Appear to begin with a protocol (`^[^\s]+\:\S`).
```
Comment (clouserw) [inline docs/metrics.md:61]: having the actual regex is great!

File: docs/metrics.md [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+![](images/kpi_1.png)
+
+
+#### User retention
+
+_Do users continue to use recommendations? As we improve the recommendations, are they more likely to stick around?_
+
+This will allow us to understand overall retention. A user is considered retained if they use at least one recommendation each week.
+
+![](images/kpi_2.png)
```
Comment (clouserw) [inline docs/metrics.md:93]: These images really help. Nice work on them

**Comment by clouserw** [comment]
aside from the JSON question, I think it looks great. thanks chuck

File: docs/metrics.md (hunk 7 above)
Comment (chuckharmston) [inline docs/metrics.md:41]: Yep, Nick requested the same thing. In this PR, I'm just talking about the plan; the actual JSON blob will come with implementation when mozilla/testpilot#479 lands and I have actual APIs to develop against.

**Comment by jaredhirsch** [comment]
@chuckharmston :tada: great stuff, let's document!
